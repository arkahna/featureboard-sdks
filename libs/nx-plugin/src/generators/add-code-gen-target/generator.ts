import type { TargetConfiguration, Tree } from '@nx/devkit'
import {
    readProjectConfiguration,
    updateProjectConfiguration,
} from '@nx/devkit'
import type { CodeGenExecutorSchema } from '../../executors/code-gen/schema'
import type { AddCodeGenTargetGeneratorSchema } from './schema'

export async function addCodeGenTargetGenerator(
    tree: Tree,
    options: AddCodeGenTargetGeneratorSchema,
) {
    const project = readProjectConfiguration(tree, options.projectName)
    project.targets ??= {}

    project.targets[options.targetName ?? 'code-gen'] = {
        ...getCodeGenExecutorConfiguration(options),
    }

    updateProjectConfiguration(tree, options.projectName, project)
}

/**
 * Returns a TargetConfiguration for the @featureboard/nx-plugin:code-gen executor
 */
export function getCodeGenExecutorConfiguration(
    options: CodeGenExecutorSchema,
): CodeGenExecutorConfiguration {
    const codeGenOptions: CodeGenExecutorConfiguration['options'] = {
        template: options.template,
        featureBoardProductName: options.featureBoardProductName,
        subFolder: options.subFolder,
    }
    if (options.dryRun) {
        codeGenOptions.dryRun = true
    }
    return {
        executor: '@featureboard/nx-plugin:code-gen',
        outputs: [`{projectRoot}/${options.subFolder}`.replace('/./', '')],
        options: codeGenOptions,
        dependsOn: ['build'],
    }
}

/**
 * Configuration options relevant for the executor
 */
export type CodeGenExecutorConfiguration = TargetConfiguration & {
    executor: '@featureboard/nx-plugin:code-gen'
    options: CodeGenExecutorSchema & { dryRun?: boolean }
}

export default addCodeGenTargetGenerator
