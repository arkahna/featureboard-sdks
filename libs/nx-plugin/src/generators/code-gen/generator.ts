import { codeGenerator } from '@featureboard/code-generator'
import type { Tree } from '@nx/devkit'
import { joinPathFragments, readProjectConfiguration } from '@nx/devkit'
import prompts from 'prompts'
import { isDryRun } from '../../Shared/is-dry-run'
import type { CodeGenGeneratorSchema } from './schema'

export async function codeGenGenerator(
    tree: Tree,
    { projectName, ...options }: CodeGenGeneratorSchema,
): Promise<void> {
    const project = readProjectConfiguration(tree, projectName)
    const dryRun = isDryRun()

    let featureBoardBearerToken: string | undefined
    if (!dryRun && !options.featureBoardKey) {
        const result = await prompts([
            {
                type: 'password',
                name: 'bearerToken',
                message: `Enter your FeatureBoard bearer token:`,
                validate: (x) => !!x,
            },
        ])
        featureBoardBearerToken = result.bearerToken
    }

    await codeGenerator({
        tree: tree,
        relativeFilePath: joinPathFragments(project.root, options.subFolder),
        interactive: !dryRun,
        featureBoardBearerToken,
        ...options,
    })
}

export default codeGenGenerator
