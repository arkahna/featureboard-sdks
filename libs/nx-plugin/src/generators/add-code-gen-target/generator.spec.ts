import type { Tree } from '@nx/devkit'
import { addProjectConfiguration, readProjectConfiguration } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'

import { addCodeGenTargetGenerator } from './generator'
import type { AddCodeGenTargetGeneratorSchema } from './schema'

describe('AddCodeGenTarget generator', () => {
    let tree: Tree
    let options: AddCodeGenTargetGeneratorSchema

    beforeEach(() => {
        tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' })

        options = {
            projectName: 'my-app',
            template: 'dotnet-api',
            featureBoardProductName: 'SaaSy Icons',
            targetName: 'saasy-icons',
            subFolder: './src/features',
            dryRun: false,
        }

        addProjectConfiguration(tree, options.projectName, {
            root: `apps/${options.projectName}`,
            projectType: 'application',
            targets: { 'my-target': { executor: 'nx:noop' } },
        })
    })

    it('should run successfully', async () => {
        await addCodeGenTargetGenerator(tree, options)
        const project = readProjectConfiguration(tree, options.projectName)
        expect(project).toBeDefined()
    })

    it('should update the project.json', async () => {
        await addCodeGenTargetGenerator(tree, options)

        const projectJson = tree
            .listChanges()
            .find((x) => x.path.endsWith('project.json'))
        expect(projectJson).toBeDefined()
        expect(projectJson?.content?.toString('utf-8')).toMatchSnapshot(
            projectJson?.path,
        )
    })
})
