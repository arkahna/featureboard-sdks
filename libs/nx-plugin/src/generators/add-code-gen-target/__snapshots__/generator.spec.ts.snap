// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddCodeGenTarget generator > should update the project.json > apps/my-app/project.json 1`] = `
"{
  \\"name\\": \\"my-app\\",
  \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
  \\"projectType\\": \\"application\\",
  \\"targets\\": {
    \\"my-target\\": {
      \\"executor\\": \\"nx:noop\\"
    },
    \\"saasy-icons\\": {
      \\"executor\\": \\"@featureboard/nx-plugin:code-gen\\",
      \\"outputs\\": [
        \\"{projectRoot}/./src/features\\"
      ],
      \\"options\\": {
        \\"template\\": \\"dotnet-api\\",
        \\"featureBoardProductName\\": \\"SaaSy Icons\\",
        \\"subFolder\\": \\"./src/features\\"
      },
      \\"dependsOn\\": [
        \\"build\\"
      ]
    }
  }
}
"
`;
